grafana/resource_machine_learning_outlier_detector.go

```diff
@@ -0,0 +1,260 @@
+			return diags
+		}
+		return diag.FromErr(err)
+	}
+
+	d.Set("name", outlier.Name)
+	d.Set("metric", outlier.Metric)
+	d.Set("description", outlier.Description)
+	if outlier.DatasourceID != 0 {
+		d.Set("datasource_id", outlier.DatasourceID)
+	} else {
+		d.Set("datasource_id", nil)
+	}
+	if outlier.DatasourceUID != "" {
+		d.Set("datasource_uid", outlier.DatasourceUID)
+	} else {
+		d.Set("datasource_uid", nil)
+	}
+	d.Set("datasource_type", outlier.DatasourceType)
+	d.Set("query_params", outlier.QueryParams)
+	d.Set("interval", outlier.Interval)
+	d.Set("algorithm", convertToSetStructure(outlier.Algorithm))
+
+	return nil
+}
+
+func resourceMachineLearningOutlierUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*client).mlapi
+	outlier, err := makeMLOutlier(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	_, err = c.UpdateOutlierDetector(ctx, outlier)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return resourceMachineLearningOutlierRead(ctx, d, meta)
+}
+
+func resourceMachineLearningOutlierDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*client).mlapi
+	err := c.DeleteOutlierDetector(ctx, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId("")
+	return nil
+}
+
+func convertToSetStructure(al mlapi.OutlierAlgorithm) []interface{} {
+	algorithmSet := make([]interface{}, 0, 1)
+	algorithmConfigSet := make([]interface{}, 0, 1)
+
+	if al.Config != nil {
+		config := map[string]interface{}{
+			"epsilon": al.Config.Epsilon,
+		}
+		algorithmConfigSet = append(algorithmConfigSet, config)
+	}
+
+	algorithm := map[string]interface{}{
+		"name":        al.Name,
+		"sensitivity": al.Sensitivity,
+		"config":      algorithmConfigSet,
+	}
+	algorithmSet = append(algorithmSet, algorithm)
+	return algorithmSet
+}
+
+func makeMLOutlier(d *schema.ResourceData, meta interface{}) (mlapi.OutlierDetector, error) {
+	datasourceID := uint(d.Get("datasource_id").(int))
+	datasourceUID := d.Get("datasource_uid").(string)
+	if datasourceID == 0 && datasourceUID == "" {
+		return mlapi.OutlierDetector{}, fmt.Errorf("either datasource_id or datasource_uid must be set")
+	}
+
+	alSet := d.Get("algorithm").(*schema.Set)
+	if alSet.Len() != 1 {
+		return mlapi.OutlierDetector{}, fmt.Errorf("at most one \"algorithm\" block is required")
+	}
+
+	al := alSet.List()[0].(map[string]interface{})
+
+	var algorithm mlapi.OutlierAlgorithm
+	algorithm.Name = strings.ToLower(al["name"].(string))
+	algorithm.Sensitivity = al["sensitivity"].(float64)
+
+	if algorithm.Name == "dbscan" {
+		config := new(mlapi.OutlierAlgorithmConfig)
+		if configSet, ok := al["config"]; ok && configSet.(*schema.Set).Len() == 1 {
+			cfg := configSet.(*schema.Set).List()[0].(map[string]interface{})
+			config.Epsilon = cfg["epsilon"].(float64)
+		} else {
+			return mlapi.OutlierDetector{}, fmt.Errorf("DBSCAN algorithm requires a single \"config\" block")
+		}
```

Comment:
Can be handled with `MaxLen: 1` + `MinLen: 1`

Reply:
The `algorithm.config` block is required only when `algorithm.name == "dbscan"`, otherwise it is optional.

I didn't see a way to encode that logic using terraform features/validators on the `schema.Resource`. Do you know a way?

Reply:
Reason I ask is combining `MinLen: 1` with `Optional: true` causes the `MinLen` to be ignored, so the above Set size length check is still needed.